File: acmicpc/acmicpc02805.cpp
#include <iostream>

using namespace std;
int main(){
    long long int n, m;
    scanf("%lld", &n);
    scanf("%lld", &m);

    long long int left = 0;
    long long int right = 0;

    long long int trees[n];
    for (int i=0; i < n; i++){
        scanf("%lld", &trees[i]);
        if (right < trees[i]){
            right = trees[i];
        }
    }

    long long int output, target;
    output = (left + right) / 2;
    while (left <= right){
        target = 0;
        for (int i=0; i < n; i++){
            if (trees[i] - output > 0){
                target = target + (trees[i] - output);
            }
        }

        if (target >= m){
            left = output + 1;
        } else if (target < m){
            right = output - 1;
        }
        output = (left + right) / 2;
    }
    cout << output << "\n";

    return 0;
}

File: acmicpc/acmicpc02805.py
import sys

def cut(output):
    target = 0
    for tree in trees:
        if (tree - output) > 0:
            target += (tree - output)
    return target

n, m = map(int, sys.stdin.readline().split())
trees = list(map(int, sys.stdin.readline().split()))

left = 0
right = max(trees)
output = (left + right) // 2
while left <= right:
    target = cut(output)

    if target >= m:
        left = output + 1
    elif target < m:
        right = output - 1
    output = (left + right) // 2

print(output)